### ISSUE TITLE
Enhancements and feedback

### ISSUE THREAD

**Issue by abomadi** [cloudscribe/sample-membership-paywall#7]
Hey @joeaudette ,
here are some of my comments regarding this module
- Webhooks for membership events, based on an event you can trigger a webhook
- Tenancy Feature Support > Filters on features that require membership
- There is a popular payment gateways that should be integrated with such as Paypal, Authorize.net, Stripe, Braintee
also there is a standard copy and pase payment methid that they enable iframe inside the website with token to process a payment
this should be very userful, i suggest to have a starter template for integratin any payment gateway for advanced user who do development
- Solid documentation how this module is used and use cases for it
- There is remminder template for user as this module intend to manually renew their subscription however recurring subscription automaticlly will give a huge push for this module with notifications based on payment status
This is for now i will give more feedback later on

**Comment by joeaudette** [comment]
Thanks that is mostly helpful feedback.

Not sure what you mean by:
> Tenancy Feature Support > Filters on features that require membership

Filter what on what?

No features depend on membership and membership can't really know about features that might be protected by membership roles.

**Comment by abomadi** [comment]
@joeaudette i missed that , make sense !, thanks

**Comment by joeaudette** [comment]
@abomadi I think the most important thing not yet implemented is support for automatic recurring payments and renewal of membership.

Square doesn't currently have an api for this, but Stripe does have a nice api that looks feasible to use for membership, so I'm going to work on that and probably make it the default/preferred payment provider instead of Square.

I think while it would be nice to have support for other payment providers it would be very challenging to implement some common abstraction to plugin other payment providers given different capabilities per provider. I'm not planning to tackle that soon and hope that if I implement Stripe with automatic renewals it will meet most people's needs. Stripe has no up front costs and no monthly costs, it only has reasonable per transaction fees, so the barrier to setup a Stripe account is very low and it seems currently to be the trending most popular solution for web payments. [Stripe is currently supported in 25 countries and growing.](https://stripe.com/global) If the membership product becomes popular and there is demand for other specific payment providers I will consider those on a case by case basis in the future.

Webhooks is not yet supported for asp.net core but I understand it is coming in the future, so any work on that will wait until then.

As far as use cases, pretty much anything that can be protected by roles could be a use case. The primary use case of course would be protected content of any kind like news etc. 

One interesting idea for a use case that has come up in discussions with colleagues and could lead to a related product later on is that it could be used for paid support contracts. For example a help desk or help ticket system where perhaps users can only enter issues if they have a member role, and the member level role could be used to prioritize issues for response times. Maybe any authenticated user could browse and search issues to reduce overall support burden by making it possible to lookup existing issues that were resolved and to help avoid duplicate issues coming up over and over. The membership system would handle the billing and access to paid support.

**Comment by abomadi** [comment]
@joeaudette i agree with the most, however regarding the payment gateway i agree with stripe however there are many countries that are not supported for now, but its a good idea to implement as its booming, but there is another approach that is commonly used called CopyAndPay which is used for fast commerce integrations with payment gateways without have any hard requirements to implement, please have a look into this https://test.ctpe.net/frontend/Integrationguide/v4/copyandpay.html

if you are willing to implement another payment gateway i suggest to have the invoicing at cloudscribe too.

**Comment by joeaudette** [comment]
@abomadi 

I've implemented Stripe and for now removed Square. If Square ever gets a good api for automatic subscription renewal I may consider implementing that.

But I'm not going to implement any new payment methods until I have some sales and an indication that this product will succeed. I think the Stripe integration should be enough to find out if there is demand for this product. If it doesn't sell using Stripe I have no reason to think adding more payment methods will make a big difference. If it sells ok with Stripe then I would see it as worth the effort to add more options.

For now I need to make videos, documentation/marketing info and then I need to move on to some other products that need my effort.
